flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/JobMaster.java

```diff
@@ -467,6 +487,128 @@ public void registerAtResourceManager(final String address) {
 		//TODO:: register at the RM
 	}
 
+	@RpcMethod
+	public NextInputSplit requestNextInputSplit(final JobVertexID vertexID, final ExecutionAttemptID executionAttempt) {
```

Comment:
I think both approaches, sending an exception or sending a failure response message in case of an exceptionally state on the receiving side, are fine. 
However, I don't agree that exceptions are only reserved for the rpc layer. "User" code should also be allowed to throw exceptions. I agree, though, that we have to pay special attention to these cases on the caller-side, because there is no compile time check that we actually handle all occurring exceptions.
All rpc related exceptions are wrapped in a `RpcConnectionException` and should, thus, be distinguishable from the "user" code exceptions.
In this specific case, I would go with throwing exceptions, because they transmit more information about the failure reason compared to simply sending `null` to the caller.